ssg/boolean_expression.py

```diff
@@ -8,18 +12,22 @@
 
 class Function(boolean.Function):
     """
-    Base class for boolean functions
-
-    Subclass it and pass to the `Algebra` as `function_cls` to enrich
-    expression elements with domain-specific methods.
-
-    Provides `is_and`, `is_or` and `is_not` methods to distinguish instances
-    between different boolean functions.
-
-    The `as_id` method will generate a unique string identifier usable as
-    an XML id based on the properties of the entity.
+    Base class for boolean functions.
+
+    This class should be subclassed and passed to the `Algebra` as `function_cls`
+    to enrich expression elements with domain-specific methods.
+
+    Methods:
+        is_and() -> bool:
```

Comment:
@marcusburghardt  just to clarify, are you sure it is a good idea to consolidate documentation for all class members here? Wouldn't it make problems when someone wants to see help for particular method?

Reply:
Keep the list of methods up-to-date is once concern I have about listing the method in the class docstring,

Reply:
I would like to move forward with this. Considering at least one case where methods mentioned in class Docstrings is useful for developers and the risk of misalignment between Docstrings and methods along the time, I would prefer to keep this for now since they are already there.

We can mitigate the risk of misalignment by exercising due diligence both when developing and when reviewing code. If this becomes a real problem in the future, we can always reevaluate and perhaps remove method information from class Docstrings.

Reply:
Hello,
I must say I am not in favor of this. I see the benefit of having everything at the one place, so that when you open documentation for the class, you see methods as well. However, I have following concerns:
1. There are specific facilities exactly for this, without writing it in the code directly. For example, VS Code has the "outline" view, see the screenshot. It is very configurable I would say.
2. It is not standard.
3. That means, that people are not used to that and we will have to keep extra eye on consistency between the real method definition and the definition in the class documentation.

In case you stand strongly behind this desire, I suggest you propose a separate PR adding this documentation to classes together with update of our style guide.

Reply:
Adding the screenshot here.
![Snímek obrazovky pořízený 2024-11-15 09-32-36](https://github.com/user-attachments/assets/992ae218-e973-48cc-9d71-14ff038002e5)

Reply:
This example of "outline" is similar to what I mentioned before, regarding `autocomplete`. The same information in your "outline" example can be obtained in different ways, but not as straightforward as in Docstrings.

In my assessment, having this offers more pros than cons. I don't believe that in this case we should remove this helpful information only because there is an hypothetical risk of having something outdated along the time, which is something we can pretty much mitigate. Also, this SSG package is relatively stable if we look the changes along the last years. In addition, I saw multiple cases of Docstrings not mentioning all methods in different projects and I found the methods anyways with some extra steps. It can happen, but what is the real problem of not having a class Docstring perfectly updated about a new method included or considerably changed, specially if the methods have their own Docstrings, which are more complete?

I don't see an strong argument against having this information there for who wants to use them among other options to obtain the same information. The intention of the documentation is to make ssg easier to be consumed and I am in favor of giving options for different preferences, specially in this case where no code is impacted.

But I want to move forward with the documentation. So, let me know if this is blocking the PR to be merged. If so, I will work on an additional commit to remove the information. If you believe this should be removed, I am fine to remove and unblock it. In any case, thanks for the feedback on this.

Reply:
I admit that having this section in the class documentation is handy, because it is just next to other information about the class, but I am still not convinced that pros outweight cons. As I am not the only one having concerns around this change, please propose it as a separate PR, preferably with an ammendment to the style guide.

Reply:
Removed in my last commit. I am personally in favor of including documentation that is helpful for some people even if not everybody is expected to use the same information, so we can support a wider audience. Specially when the information causes no harm to the project. But I removed the information so we can move forward with this PR and discuss preferences in another place. : )